### PR TITLE
Create documentation on debugging on mac

### PR DESCRIPTION
**Pull request by Forgind**
Add documentation for how to run a single unit test and see the outputs of Console.WriteLine statements on a Mac.

### CODE REVIEW

File: documentation/wiki/Mac-Debugging.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#Debugging with MacOS
+Say unit tests are failing from Microsoft.Build.Engine.UnitTests_netcoreapp2.1_x64
```
Comment (rainersigwald) [inline documentation/wiki/Mac-Debugging.md]: Can you include some console output from `build.sh -test` showing how to know that this is the thing that's failing?

File: documentation/wiki/Mac-Debugging.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+* Open terminal 
+  * Hit command-space, type terminal, hit enter. Alternatively, hit magnifying glass (spotlight) in upper-right corner and search for terminal.)
+* Build something
+  * Navigation in terminal is similar to command prompt (cd), although you type `ls` in place of `dir`.
+  * **Use ./build.sh instead of .\build.cmd.**
```
Comment (rainersigwald) [inline documentation/wiki/Mac-Debugging.md]: Do you think it's valuable to have this level of OS documentation here? We tend to assume command-line skills and focus on what's unique(ish) to this repo.

File: documentation/wiki/Mac-Debugging.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+* Type `find . -name Microsoft.Build.Engine.UnitTests_netcoreapp2.1_x64.log`
```
Comment (rainersigwald) [inline documentation/wiki/Mac-Debugging.md]: I'd emphasize the `.html` -> `.log` transformation from the build error.

File: documentation/wiki/Mac-Debugging.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  * This should print out a path (from your current working directory) to the relevant log file.
+* Type `tail <path from previous step>`
+  * This prints out the last part of that file.
+  * You can also just open it normally from a finder window.
+  * This file contains the standard output from the last run.
+  * You may notice that the file ends with `=== COMMAND LINE ===` followed by a single (long) command line statement.
```
Comment (rainersigwald) [inline documentation/wiki/Mac-Debugging.md]: If you take the `tail -1` edit above, the command line will be the only output so this becomes a factoid about opening the whole file (still valuable) instead of critical.

File: documentation/wiki/Mac-Debugging.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+* Copy the command line statement from the previous step. Remove the portion after the redirection (`>` character). You may notice that part redirects output to the file you’re viewing.
+  * The last part (`2>&1`) redirects standard error (using `2>`) to the same place as where standard out is going (`&1`), in this case this log file.
+* Append `./build.sh &&` to the beginning of the truncated command line statement.
```
Comment (rainersigwald) [inline documentation/wiki/Mac-Debugging.md]:
```suggestion
* Prepend `./build.sh &&` to the beginning of the truncated command line statement.
```

File: documentation/wiki/Mac-Debugging.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+* Append `-method ` and the name of the method you want to test to the end.
```
Comment (rainersigwald) [inline documentation/wiki/Mac-Debugging.md]: add something about getting the method from opening the HTML log in a browser?

File: documentation/wiki/Mac-Debugging.md (hunk 5 above)
Comment (rainersigwald) [inline documentation/wiki/Mac-Debugging.md]: I'd separate the "inner loop iterating on a test with `build && runtest` from the "running a single test in isolation" concepts--it's probably more valuable to doc the former but it's easier to understand the latter.

File: documentation/wiki/Mac-Debugging.md (hunk 6 above)
Comment (rainersigwald) [inline documentation/wiki/Mac-Debugging.md]: I wanted to say "link to the docs on how to do filters" but I can't find any such docs :(

Probably worth mentioning `-class` to run all tests in a class, and that the name provided must be fully qualified (including namespace).

File: documentation/wiki/Mac-Debugging.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+*	Running this statement will run just the one test (after building) and print out both the error and the Console.WriteLine() statements you added to the test/what it calls.
+
+
+Sample statements with outputs below (*italicized*) and changes to the output of the second command **bolded** (note that additionally, the last two lines and one character of the second output were deleted):
```
Comment (rainersigwald) [inline documentation/wiki/Mac-Debugging.md]: For things like this I like to use the `sh-session` syntax highlighting, like so:

```sh-session
$ find . -name Microsoft.Build.Engine.UnitTests_netcoreapp2.1_x64.log
./artifacts/log/Debug/Microsoft.Build.Engine.UnitTests_netcoreapp2.1_x64.log

$ tail ./artifacts/log/Debug/Microsoft.Build.Engine.UnitTests_netcoreapp2.1_x64.log
AnyEvent:message
CustomWarning:Event type "MyCustomBuildWarningEventArgsNotSerializable" was expected to be serializable using the .NET serializer. The event was not serializable and has been ignored.
AnyEvent:Event type "MyCustomBuildWarningEventArgsNotSerializable" was expected to be serializable using the .NET serializer. The event was not serializable and has been ignored.
CustomWarning:Event type "MyCustomBuildErrorEventArgsNotSerializable" was expected to be serializable using the .NET serializer. The event was not serializable and has been ignored.
AnyEvent:Event type "MyCustomBuildErrorEventArgsNotSerializable" was expected to be serializable using the .NET serializer. The event was not serializable and has been ignored.
  Finished:    Microsoft.Build.Engine.UnitTests
=== TEST EXECUTION SUMMARY ===
   Microsoft.Build.Engine.UnitTests  Total: 2928, Errors: 0, Failed: 0, Skipped: 47, Time: 112.394s
=== COMMAND LINE ===
"/Users/nathanmytelka/Desktop/code/msbuild/.dotnet/dotnet" exec --depsfile "/Users/nathanmytelka/Desktop/code/msbuild/artifacts/bin/Microsoft.Build.Engine.UnitTests/Debug/netcoreapp2.1/Microsoft.Build.Engine.UnitTests.deps.json" --runtimeconfig "/Users/nathanmytelka/Desktop/code/msbuild/artifacts/bin/Microsoft.Build.Engine.UnitTests/Debug/netcoreapp2.1/Microsoft.Build.Engine.UnitTests.runtimeconfig.json"  "/Users/nathanmytelka/.nuget/packages/xunit.runner.console/2.4.1/tools/netcoreapp2.0/xunit.console.dll" "/Users/nathanmytelka/Desktop/code/msbuild/artifacts/bin/Microsoft.Build.Engine.UnitTests/Debug/netcoreapp2.1/Microsoft.Build.Engine.UnitTests.dll" -noautoreporters -xml "/Users/nathanmytelka/Desktop/code/msbuild/artifacts/TestResults/Debug/Microsoft.Build.Engine.UnitTests_netcoreapp2.1_x64.xml" -html "/Users/nathanmytelka/Desktop/code/msbuild/artifacts/TestResults/Debug/Microsoft.Build.Engine.UnitTests_netcoreapp2.1_x64.html" -notrait category=nonosxtests -notrait category=netcore-osx-failing -notrait category=nonnetcoreapptests -notrait category=failing > "/Users/nathanmytelka/Desktop/code/msbuild/artifacts/log/Debug/Microsoft.Build.Engine.UnitTests_netcoreapp2.1_x64.log" 2>&1
$ ./build.sh && "/Users/nathanmytelka/Desktop/code/msbuild/.dotnet/dotnet" exec --depsfile "/Users/nathanmytelka/Desktop/code/msbuild/artifacts/bin/Microsoft.Build.Engine.UnitTests/Debug/netcoreapp2.1/Microsoft.Build.Engine.UnitTests.deps.json" --runtimeconfig "/Users/nathanmytelka/Desktop/code/msbuild/artifacts/bin/Microsoft.Build.Engine.UnitTests/Debug/netcoreapp2.1/Microsoft.Build.Engine.UnitTests.runtimeconfig.json"  "/Users/nathanmytelka/.nuget/packages/xunit.runner.console/2.4.1/tools/netcoreapp2.0/xunit.console.dll" "/Users/nathanmytelka/Desktop/code/msbuild/artifacts/bin/Microsoft.Build.Engine.UnitTests/Debug/netcoreapp2.1/Microsoft.Build.Engine.UnitTests.dll" -noautoreporters -xml "/Users/nathanmytelka/Desktop/code/msbuild/artifacts/TestResults/Debug/Microsoft.Build.Engine.UnitTests_netcoreapp2.1_x64.xml" -html "/Users/nathanmytelka/Desktop/code/msbuild/artifacts/TestResults/Debug/Microsoft.Build.Engine.UnitTests_netcoreapp2.1_x64.html" -notrait category=nonosxtests -notrait category=netcore-osx-failing -notrait category=nonnetcoreapptests -notrait category=failing **-method Microsoft.Build.UnitTests.BackEnd.TaskBuilder_Tests.NullMetadataOnLegacyOutputItems
```

Unfortunately, it doesn't allow bolding parts of command lines, and it looks like it's misclassifying the command-line-from-tailing-the-log as an entered command line, which is unfortunate, so maybe stick with the current approach for this one. But please wrap in `<pre>` and use `<i>` and `<b>`; proportional computer output hurts my heart.

File: documentation/wiki/Mac-Debugging.md (hunk 2 above)
Comment (Forgind) [reply documentation/wiki/Mac-Debugging.md]: I do, primarily because the audience for this document includes people who have lived their entire lives in the Microsoft ecosystem. I can imagine these sorts of small differences derailing someone for fifteen minutes as they look for something very specific that doesn't exist on Macs. I don't think it will really bother anyone.

File: documentation/wiki/Mac-Debugging.md (hunk 4 above)
Comment (Forgind) [reply documentation/wiki/Mac-Debugging.md]: That doesn't seem to be an option?

Nathans-MacBook-Pro:msbuild nathanmytelka$ tail -l ./artifacts/log/Debug/Microsoft.Build.Engine.UnitTests_netcoreapp2.1_x64.log
tail: illegal option -- l
usage: tail [-F | -f | -r] [-q] [-b # | -c # | -n #] [file ...]

File: documentation/wiki/Mac-Debugging.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
+#Debugging with MacOS
+Say unit tests are failing from Microsoft.Build.Engine.UnitTests_netcoreapp2.1_x64
+* Open terminal 
+  * Hit command-space, type terminal, hit enter. Alternatively, hit magnifying glass (spotlight) in upper-right corner and search for terminal.)
+* Build something
+  * Navigation in terminal is similar to command prompt (cd), although you type `ls` in place of `dir`.
+  * **Use ./build.sh instead of .\build.cmd.**
+* Type `find . -name Microsoft.Build.Engine.UnitTests_netcoreapp2.1_x64.log`
+  * This should print out a path (from your current working directory) to the relevant log file.
+* Type `tail <path from previous step>`
```
[the remark this answers is not in the document]
Comment (rainersigwald) [reply documentation/wiki/Mac-Debugging.md]: Looks like BSD `tail` in macOS doesn't take GNU `tail`'s shortcut option for this. I think `-n 1` will work in both variants.

File: documentation/wiki/Mac-Debugging.md (hunk 5 above)
Comment (Forgind) [reply documentation/wiki/Mac-Debugging.md]: Do you mean that you'd separate the building step from the running test step? If so, that might confuse people who see a bug, fix the bug, run the test again, and get the same bug because they didn't rebuild.

File: documentation/wiki/Mac-Debugging.md (hunk 6 above)
Comment (Forgind) [reply documentation/wiki/Mac-Debugging.md]: I tried ./build.sh -class..., and it told me that ./build.sh has no -class switch. Did you mean adding -class to something else?

Comment (rainersigwald) [reply documentation/wiki/Mac-Debugging.md]: yes, to the test command line

Comment (Forgind) [reply documentation/wiki/Mac-Debugging.md]: I'm not sure in what situations that would be helpful. If you know what class, then you presumably ran all tests at some point, and if you ran all tests, you can look at the html file and find which methods specifically, in which case you don't need to run a class full of tests. If you already know that a particular method failed, and you want to know if your fix fixed it, you'll want to test the method specifically, not the whole class. The log from stdout is a mess, and although it's possible, it's difficult to figure out what actually went wrong. stderr is better, but it's still less helpful than the html file because it doesn't really say why a test failed. In summary, what's the use case?

Comment (rainersigwald) [reply documentation/wiki/Mac-Debugging.md]: I use `-class` probably 1/4 of the time that I run tests on the command line.

It's particularly useful when refactoring product code in a particular area: if it breaks something, it's likely to be in "the tests that cover this stuff", as opposed to entirely unrelated tests. But you don't know exactly _what_ might break.

It's also nice when you're changing a test utility method used in multiple test cases, or if multiple test cases failed for closely related reasons.

You certainly could list all of the failing tests individually, but I generally feel that if you're going to run more than one test, it's not worth scoping it to the minimum subset of interest.

File: documentation/wiki/Mac-Debugging.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+* Type `tail -n 1 <path from previous step>`
+  * This prints out the last line of that file.
+  * You can also just open it normally from a finder window.
+  * This file contains the standard output from the last run.
+  * You may notice that the line printed by this command is a single (long) command line statement.
+* Copy the command line statement from the previous step. Remove the portion after the redirection (`>` character not preceded by 2) including that character. You may notice that part redirects output to the file you’re viewing.
+  * The last part (`2>&1`) redirects standard error (using `2>`) to the same place as where standard out is going (`&1`), in this case this log file.
+  * If you would like to rerun all tests from a given class (rather than just a specific method), you can append `-class` and the class's fully qualified name. To run all tests from `TaskBuilder_Tests`, for instance, you would add `-class Microsoft.Build.UnitTests.BackEnd.TaskBuilder_Tests` and run the statement without the following steps. Note that in the example below, standard output is redirected to `/dev/null`, thus only printing the errors.
```
Comment (rainersigwald) [inline documentation/wiki/Mac-Debugging.md]: Why pipe to `/dev/null`? Doesn't that swallow the summary and the diagnostic output of well-behaved tests?

Comment (Forgind) [reply documentation/wiki/Mac-Debugging.md]: It creates a lot of hard-to-read output. If you're using this doc, you presumably aren't familiar enough with MSBuild/macs to understand everything it outputs. This way, since stderr isn't redirected, it only shows you what you actually need.
